### PR TITLE
[FIX] base: sum of company-dependent fields in partner merge wizard

### PR DESCRIPTION
**Pull request by admi2806**
How to reproduce the bug:

- Install POS and data_cleaning
- Create 2 customers and set loyalty points
- Select both customers and merge them
- Once the merge is done, the loyalty points have not been summed

Bug:

When you merge 2 contacts that have both loyalty points, the resulting
contact won't have the sum of the 2 source contacts. Instead of that,
the resulting contact will only have the points of one of of the
contact.

opw: 2686208

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by admi2806** [comment]
@simongoffin 
Can you please have a look at this plz ?

**Review by simongoffin — COMMENTED** [review]
Hello J'ai vu que le champ loyalty_points est company_dependent. Que se passe-t-il en multi compagnie? Ex Considérons deux compagnies C1, C2 et deux partenaires P1, P2
P1 a 3 points dans C1 et 2 points dans C2
P2 a 4 points dans C1 et 1 point dans C2
Si on merge P1 et P2 en P3
P3 aura t il bien 7 points dans C1 et 3 points dans C2?

**Comment by admi2806** [comment]
> Hello J'ai vu que le champ loyalty_points est company_dependent. Que se passe-t-il en multi compagnie? Ex Considérons deux compagnies C1, C2 et deux partenaires P1, P2 P1 a 3 points dans C1 et 2 points dans C2 P2 a 4 points dans C1 et 1 point dans C2 Si on merge P1 et P2 en P3 P3 aura t il bien 7 points dans C1 et 3 points dans C2?

Bien vu, je vais régler ça :)

**Comment by admi2806** [comment]
> Hello J'ai vu que le champ loyalty_points est company_dependent. Que se passe-t-il en multi compagnie? Ex Considérons deux compagnies C1, C2 et deux partenaires P1, P2 P1 a 3 points dans C1 et 2 points dans C2 P2 a 4 points dans C1 et 1 point dans C2 Si on merge P1 et P2 en P3 P3 aura t il bien 7 points dans C1 et 3 points dans C2?

Hello, après une discussion avec le Support Expert le PO de POS, Il faut bien gérer ce cas-là. Je vais le modifier maintenant du coup :)

**Review by simongoffin — APPROVED** [review]

**Review by nle-odoo — COMMENTED** [review]
I'm not sure it's correct to just get  self.env.user.company_ids since it's restricting to current user companies, but there could be data in other companies

I guess if we need to have them, you would need to check on ir.property directly, something like:

```py
field_id = self.env['ir.model.fields']._get('res.partner', column).id 
company_ids = self.env['ir.property'].sudo().search([
    ('fields_id', '=', field_id),
    ('res_id', 'in', src_partners.ids + [dst_partner_id.id]),
    ('company_id', '!=', False)
]).mapped('company_id')
```

I'm not sure how right that would be, but if you are getting the properties maybe you could maybe just do something like this:

```py
for property in properties:
        company_values = values_company_dependent.setdefault(property.company_id, {})
        if company_values.get(column):
            company_values[column] += property.get_by_record()
        elif property.get_by_record():
            company_values[column] += property.get_by_record()
```

**Comment by admi2806** [comment]
> 

Hello,
When I wrote the patch, I was not sure that only getting the current user companies would be enough but I'm pretty sure that if you try to access these values while the company is not active, you will get a security access error.

**Comment by simongoffin** [comment]
@admi2806 Can you add a test for it?

**Comment by admi2806** [comment]
> @admi2806 Can you add a test for it?

I have added a test in the pos_loyalty module which is part of enterprise.
Here is a link to the PR: https://github.com/odoo/enterprise/pull/24059

File: odoo/addons/base/wizard/base_partner_merge.py [hunk 1]
```diff
@@ -244,6 +244,7 @@ def _update_values(self, src_partners, dst_partner):
 
         model_fields = dst_partner.fields_get().keys()
         summable_fields = self._get_summable_fields()
+        res_id = ['res.partner,{}'.format(id) for id in src_partners.ids + [dst_partner.id]]
```
Comment (rco-odoo) [inline odoo/addons/base/wizard/base_partner_merge.py:247]: This is a list, please use some consistent naming like `res_ids`.

File: odoo/addons/base/wizard/base_partner_merge.py [hunk 2]
```diff
@@ -252,6 +253,7 @@ def write_serializer(item):
                 return item
         # get all fields that are not computed or x2many
         values = dict()
+        values_company_dependent = dict()
```
Comment (rco-odoo) [inline odoo/addons/base/wizard/base_partner_merge.py:256]: Use a variable name that can be read and understood, and a `defaultdict` will make code easier.
```suggestion
        values_by_company = defaultdict(lambda: defaultdict(int))   # {company_id: {fname: value}}
```

File: odoo/addons/base/wizard/base_partner_merge.py [hunk 3]
```diff
@@ -261,10 +263,28 @@ def write_serializer(item):
...
+                                ]).mapped('company_id')
+                for company_id in company_ids:
+                    if company_id not in values_company_dependent:
+                        values_company_dependent[company_id] = dict()
+                    for item in itertools.chain(src_partners, [dst_partner]):
+                        if item.with_company(company_id)[column]:
+                            if column in values_company_dependent[company_id]:
+                                values_company_dependent[company_id][column] += write_serializer(item.with_company(company_id)[column])
+                            else:
+                                values_company_dependent[company_id][column] = write_serializer(item.with_company(company_id)[column])
```
Comment (rco-odoo) [inline odoo/addons/base/wizard/base_partner_merge.py:281]: Using `ir.property`'s model details is a big mistake IMO.
I suggest iterating on _all_ companies, and use `sudo()` (it is reasonable in this case).
With the `defaultdict` above, it becomes
```suggestion
                for company in self['res.company'].sudo().search([]):
                    for partner in (src_partners + dst_partner).sudo().with_company(company):
                        values_by_company[company][column] += write_serializer(partner[column])
```

File: odoo/addons/base/wizard/base_partner_merge.py [hunk 4]
```diff
@@ -261,10 +263,28 @@ def write_serializer(item):
...
         # remove fields that can not be updated (id and parent_id)
         values.pop('id', None)
         parent_id = values.pop('parent_id', None)
         dst_partner.write(values)
+        for company_id in values_company_dependent:
+            dst_partner.with_company(company_id).write(values_company_dependent[company_id])
```
Comment (rco-odoo) [inline odoo/addons/base/wizard/base_partner_merge.py:287]: If you do that on all companies, you should write with a sudo(), which looks fine to me in this context.

**Review by rco-odoo — CHANGES_REQUESTED** [review]
I second the remark above: you should consider _all_ companies, as the result of the operation should not depend on the current user.

**Comment by admi2806** [comment]
Hello @rco-odoo ,
Thanks for the review. I have taken your comments into account and your solution is the best one.

**Review by nle-odoo — APPROVED** [review]
I was afraid of doing `self.env['res.company'].sudo().search([])`, but looks a lot clearer like that :+1: 

looks good to me after removing:

```py
res_id = ['res.partner,{}'.format(id) for id in src_partners.ids + [dst_partner.id]]
```
that is no longer used

and commit message should  have module `base` in title

**Comment by rco-odoo** [comment]
@admi2806 I modified a bit the commit:
- changed the commit's title
- removed useless `res_id`
- using `sum()` instead of a loop makes the computation more readable

**Review by rco-odoo — APPROVED** [review]

**Comment by rco-odoo** [comment]
@simongoffin @nle-odoo I let you review my last changes, and merge the pull request.

**Review by simongoffin — APPROVED** [review]
Ok it looks good

**Comment by simongoffin** [comment]
robodoo r+

**Pull request closed without merging**